### DISCUSSION TITLE
reasons for slow upload?

### DISCUSSION THREAD

**Discussion by dankeyy** [pocketbase/pocketbase#3011] [Q&A] [answered]
Running 100% locally, no network latency at all, 
it takes 2.5-3 seconds to upload a 0.5GB file to the server - an out of the box pocketbase binary from here https://pocketbase.io/docs/ (both linux and macos)
on an SSD with write speeds that should easily handle a size like that in a fraction of the time

what could possibly cause this?

**Comment by ganigeorgiev** [comment, marked as answer]
In general you'll never get the same write speeds as operating directly with the file on the disk.

Note that although you are on localhost, the client/browser still has to prepare the file to send it over the HTTP connection.

Additionally, since the file is larger than 32MB, the request data will be not kept entirely in memory and a new temporary file on the disk will be created, so I expect some copying/allocations to be involved.

We also have a filesystem abstraction (based on [go-cloud/blob](https://github.com/google/go-cloud/tree/master/blob)) in order to have a single interface for both local filesystem and S3 and we inherit some of its costs.

Another thing that could also affect the performance if the MIME type option is specified, is that we have to read the first couple bytes of the file to find its real MIME type among a list of other known signatures.

Can it be improved? Probably Yes, I have plans to simplify the fs abstraction but it is left aside for now as other tasks are with higher priority.
